plugins/uast-kotlin-fir/src/org/jetbrains/uast/kotlin/declarations/FirKotlinUMethod.kt

```diff
@@ -0,0 +1,152 @@
+import org.jetbrains.kotlin.fir.declarations.FirFunction
+import org.jetbrains.kotlin.fir.declarations.FirProperty
+import org.jetbrains.kotlin.fir.declarations.FirPropertyAccessor
+import org.jetbrains.kotlin.fir.expressions.FirBlock
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getOrBuildFirOfType
+import org.jetbrains.kotlin.idea.fir.low.level.api.api.getResolveState
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
+import org.jetbrains.kotlin.utils.SmartList
+import org.jetbrains.uast.*
+
+open class FirKotlinUMethod(
+    psi: PsiMethod,
+    final override val sourcePsi: KtDeclaration?,
+    givenParent: UElement?
+) : FirKotlinAbstractUElement(givenParent), UMethod, UAnchorOwner, PsiMethod by psi {
+    constructor(psi: KtLightMethod, givenParent: UElement?) : this(psi, getKotlinMemberOrigin(psi), givenParent)
+
+    override val psi: PsiMethod = unwrap<UMethod, PsiMethod>(psi)
+
+    override val javaPsi = psi
+
+    override fun getSourceElement() = sourcePsi ?: this
+
+    private val kotlinOrigin by lz { getKotlinMemberOrigin(psi.originalElement) ?: sourcePsi }
+
+    override val uAnnotations: List<UAnnotation>
+        get() {
+            // TODO: Not yet implemented
+            return emptyList()
+        }
+
+    private val receiverTypeReference by lz {
+        when (sourcePsi) {
+            is KtCallableDeclaration -> sourcePsi
+            is KtPropertyAccessor -> sourcePsi.property
+            else -> null
+        }?.receiverTypeReference
+    }
+
+    override val uastParameters: List<UParameter> by lz {
+
+        fun parameterOrigin(psiParameter: PsiParameter?): KtElement? {
+            return when (psiParameter) {
+                is KtLightElement<*, *> -> psiParameter.kotlinOrigin
+                // TODO: UastKotlinPsiParameter?
+                else -> null
+            }
+        }
+
+        val lightParams = psi.parameterList.parameters
+        val receiverTypeReference =
+            receiverTypeReference ?: return@lz lightParams.map { FirKotlinUParameter(it, parameterOrigin(it), this) }
+        val lightReceiver = lightParams.firstOrNull() ?: return@lz emptyList<UParameter>()
+        val uParameters = SmartList<UParameter>(FirKotlinReceiverUParameter(lightReceiver, receiverTypeReference, this))
+        lightParams.drop(1).mapTo(uParameters) { FirKotlinUParameter(it, parameterOrigin(it), this) }
+        uParameters
+    }
+
+    override val uastAnchor: UIdentifier? by lz {
+        val identifierSourcePsi = when (val sourcePsi = sourcePsi) {
+            is PsiNameIdentifierOwner -> sourcePsi.nameIdentifier
+            is KtObjectDeclaration -> sourcePsi.getObjectKeyword()
+            is KtPropertyAccessor -> sourcePsi.namePlaceholder
+            else -> sourcePsi?.navigationElement
+        }
+        FirKotlinUIdentifier(nameIdentifier, identifierSourcePsi, this)
+    }
+
+    protected val firFunction: FirFunction<*>? by lz {
```

Comment:
Note that I didn't leave a comment about this because I think `KtSymbol` world doesn't have abstractions for expressions. That is, to convert function body, this is the last place we can get FIR element. Let me know if I'm wrong.

Reply:
Can we just use KtExpression here and do not access a FIR elements at all? We can do a lot of work with it without going directly into FIR. E.g, we can resolve it if it is reference by `KtReference.resolve()`, get it's type by `KtExpressionTypeProviderMixIn.getKtType`, etc

Reply:
I wasn't aware of KtExpression, thanks!

So, until I dive into expression-level conversion, removed FIR use here.